refactor(db_engine_specs): Removing top-level import of app

john-bodley opened the pull request:
### SUMMARY

Yak shaving at its finest. This PR started due to the fact that some of the database engine specs had the following import logic,

```python
from superset import app
```

which was very problematic due to potential circular imports and app contexts which needed to be invoked. For testing we required magical logic like

```python
from tests.test_app import app
```

prior to importing the actual engine spec.

This change ensures that the engine specs are decoupled from the Flask app and use the `flask.current_app` where necessary when the application configuration is required. 

Note what started out as somewhat of a benign change quickly spiraled out of control in order to ensure that the tests passed as more and more code needed to be touched. The TL;DR is:

- ~~Even though this logic could further be improved, in my opinion the approach is in the direction of the north star, i.e., removing the dependency for `tests.test_app` and using the `app` `pytest` fixture instead.~~
- ~~The `app` fixture is scoped to the function which ensures that the tests are idempotent and thus the `app.config` neither i) needs to be mocked, or ii) reset upon completion of the test.~~
- ~~Migrates a subset of tests to be more `pytest` like so we can leverage the `app` fixture.~~ 

This logic was retracted because even though the individual tests passed they failed as part of the test suite and further debugging the issue resulted in a real nightmare.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TEST PLAN
<!--- What steps should be taken to verify the changes -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Commit:
refactor(db_engine_specs): Removing top-level import of app

villebro reviewed (approved):
This looks great! Some tests seems to be failing, but approving the changes as this is clearly the better path forward.

john-bodley commented:
@villebro any thoughts as to why CI may be failing? I've been wrangling with this for hours, descoping the impact of the change as I go. Note it seems like the test hardness is failing on `tests/queries/api_tests.py` though passes when run separately, i.e., 


```
tox -e py38-sqlite -- tests/queries/api_tests.py
```

which would possibly indicate some leakage, however I've gone over the change a number of times and can't see what the issue could be. For example I renamed the fixture I added from `app` to `test_app` to now `app_` to make sure there's no conflict with the whole `tests.test_app.app` yuckiness which gets imported (in various ways) and is used in a number of fixtures.

john-bodley commented on tests/conftest.py:
```diff
@@ -18,18 +18,27 @@
+# TODO(john-bodleY): The `tests.test_app.app` logic is atypical and extremely brittle
+# and thus should be deprecated in favor of a more typical Flask testing model utilizing
+# a function scoped app to aid with idempotent testing.
+@pytest.fixture
```
@villebro the inclusion of this fixture to help ensure idempotent testing illustrates just how fragile our testing configuration is, i.e., using the fixture is insolation works but causes other tests to fail when testing the entire suite of tests.

I think there  needs to be a saner mechanism for creating an application at the function (default) scope whilst leveraging the examples database allowing for simpler more robust idempotent testing.
